src/vfs2.c

```diff
@@ -404,128 +490,116 @@ static int check_wal_integrity(sqlite3_file *f)
+	sqlite3_free(e);
+
+}
 
 static int vfs2_close(sqlite3_file *file)
 {
-	int rv, rvprev;
-	struct vfs2_file *xfile = (struct vfs2_file *)file;
-
-	unregister_file(xfile);
+	struct file *xfile = (struct file *)file;
+	int rv;
 
-	rvprev = SQLITE_OK;
-	if (xfile->flags & SQLITE_OPEN_WAL) {
-		sqlite3_free(xfile->wal.wal_cur_fixed_name);
-		sqlite3_free(xfile->wal.wal_prev_fixed_name);
-		if (xfile->wal.wal_prev->pMethods != NULL) {
-			rvprev = xfile->wal.wal_prev->pMethods->xClose(
-			    xfile->wal.wal_prev);
-		}
-		if (xfile->wal.pending_txn_frames != NULL) {
-			for (uint32_t i = 0; i < xfile->wal.pending_txn_len;
-			     i++) {
-				sqlite3_free(
-				    xfile->wal.pending_txn_frames[i].data);
-			}
-		}
-		sqlite3_free(xfile->wal.pending_txn_frames);
-		sqlite3_free(xfile->wal.wal_prev);
-	} else if (xfile->flags & SQLITE_OPEN_MAIN_DB) {
-		for (int i = 0; i < xfile->db_shm.regions_len; i++) {
-			sqlite3_free(xfile->db_shm.regions[i]);
-		}
-		sqlite3_free(xfile->db_shm.regions);
-	}
 	rv = SQLITE_OK;
-	if (xfile->orig->pMethods != NULL) {
+	if (xfile->flags & SQLITE_OPEN_MAIN_DB) {
+		if (xfile->orig->pMethods != NULL) {
+			rv = xfile->orig->pMethods->xClose(xfile->orig);
+		}
+		sqlite3_free(xfile->orig);
+		xfile->entry->refcount_main_db -= 1;
+		maybe_close_entry(xfile->entry);
+	} else if (xfile->flags & SQLITE_OPEN_WAL) {
+		xfile->entry->refcount_wal -= 1;
+		maybe_close_entry(xfile->entry);
+	} else if (xfile->orig->pMethods != NULL) {
 		rv = xfile->orig->pMethods->xClose(xfile->orig);
+		sqlite3_free(xfile->orig);
 	}
-	sqlite3_free(xfile->orig);
-	if (rv != SQLITE_OK) {
-		return rv;
-	}
-	return rvprev;
+	return rv;
 }
 
 static int vfs2_read(sqlite3_file *file, void *buf, int amt, sqlite3_int64 ofst)
 {
-	struct vfs2_file *xfile = (struct vfs2_file *)file;
-	return xfile->orig->pMethods->xRead(xfile->orig, buf, amt, ofst);
+	struct file *xfile = (struct file *)file;
+	sqlite3_file *orig = get_orig(xfile);
+	return orig->pMethods->xRead(orig, buf, amt, ofst);
 }
 
-static int vfs2_wal_swap(struct vfs2_file *wal,
-			 const struct vfs2_wal_hdr *wal_hdr)
+static int wal_swap(struct entry *e,
+		    const struct wal_hdr *wal_hdr)
 {
-	PRE(wal->wal.pending_txn_len == 0);
-	PRE(wal->wal.pending_txn_frames == NULL);
+	PRE(e->pending_txn_len == 0);
+	PRE(e->pending_txn_frames == NULL);
 	int rv;
 
-	atomic_uint *p = &wal->vfs_data->page_size;
-	unsigned expected = 0;
-	uint32_t z = ByteGetBe32(wal_hdr->page_size);
-	if (!atomic_compare_exchange_strong(p, &expected, z)) {
-		assert(expected == z);
-	}
+	e->page_size = ByteGetBe32(wal_hdr->page_size);
 
-	sqlite3_file *phys_outgoing = wal->orig;
-	char *name_outgoing = wal->wal.wal_cur_fixed_name;
-	sqlite3_file *phys_incoming = wal->wal.wal_prev;
-	char *name_incoming = wal->wal.wal_prev_fixed_name;
+	sqlite3_file *phys_outgoing = e->wal_cur;
+	char *name_outgoing = e->wal_cur_fixed_name;
+	sqlite3_file *phys_incoming = e->wal_prev;
+	char *name_incoming = e->wal_prev_fixed_name;
```

Comment:
Hmm, I was hoping that using the `incoming` and `outgoing` terminology would be clearer since if I refer to them as `prev` and `cur` it's not clear whether those names refer to the situation just before the swap or just after. Which of those conventions would make more sense to you?

Reply:
I've added a comment about this as well